### DISCUSSION TITLE
Review of the delay feature. No more flood alerts. Problem solved.

### DISCUSSION THREAD

**Discussion by tissole** [aahnik/tgcf#132] [General]
As you know a big concern with this rapid method of forwarding is the flood alerts raised by TG that some think could lead to an account ban. There are 2 cases reported here, many rumors and even warnings from the developers of similar scripts on Github.

A feature to control the forwarding speed was requested by users, to reduce the risk. @aahnik implemented a mechanism so we could test it. He chose the delay feature. 

I created a fresh new channel and starting the test with an initial delay of 3 seconds. The script is working 1s and waits for 3s. I tested in smaller time windows of 3 minutes and then extrapolated to 1h. In 3 minutes the script has forwarded 57 msg, ~1.26 msg/s. In 1 hour will forward 1140 msg. That's too few!

Next, I lowered the delay to 2s. In 3 minutes has forwarded 83 msg, ~1.38msg/s. In 1 hour that results in 1660 msg. Still too few!

Then I tested with the lowest delay of 1s. Has forwarded 154 msg in 3 minutes, ~1.71 msg/s. In 1 hour 3080 msg will be forwarded. That's too many!

We could see that the speed increases as the delay decreases. The max speed is when there is no delay, over 3 msg/s, but that triggers flood. So I needed a delay between 1 and 2 and @aahnik was kind to implement floating numbers.

I tested for bigger timeframes -5 hours- with a delay of 1.65. After 1h 1983 msg were forwarded and no flood!!!. After 5 hours has forwarded 9873 msg, that's 1974 msg/h and no flood alerts!!! 

The next test was done with a value of 1.63. After 4 hours the script has forwarded 7989 msg, 1997/h. Still no flood alerts!!

The final test was done with a delay of 1.62s. After 5h 10015 messages were forwarded, 2003/h. No flood alerts! 

In conclusion, these values are suitable for long sessions without triggering flood alerts. Feel free to test this feature yourself and give feedback. It would be interesting to find the lowest delay that does not trigger flood and then the next value that triggers those alerts. Use lower values and longer timeframes.

Thank you @aahnik for this feature, I know it was not a top priority and you were skeptical about it, but is working. The script is as fast as before and the flood messages are gone.

**Comment by aahnik** [comment]
i am glad that the problem is solved! @tissole

**Reply by aahnik** [reply]
hi @tissole, you are an expert, so please help others in the GitHub discussions and issues. I am not getting enough time to respond to people.

**Reply by ankit-sinha-18** [reply]
@tissole I have checked with the delay also And there is no floodwaits when delay is given. I personally use delay of 3 sec always(when there were Telegram chat forward,I have added delay to prevent floodwaits) and not a single account banned due to this.

Two days ago,I suggested my friend to use tgcf for forwarding message.
I helped him to configure yml file and also set the delay of 3 sec. He had got no floodwaits but still Telegram banned his account.
Not even ,a single one ,his 7 accounts get blocked in just two days
I had personally seen his logs, there were no floodwaits but still Telegram banned his account.
There were written, account is deleted or deactivated (due to sendmessage request) but there were no floodwaits from Telegram.
Two were VOIP number and rest were genuine number of India.
He had also sent  mail to Telegram but no response from Telegram.

I have also forwarded lakhs of messages from one Channel to another Channel with delay ,but my number didn't get ban ,.

So , floodwaits may not be a single cause of account ban,there must be something else more.

**Reply by tissole** [reply]
I'm sorry for your friend. If I understand correctly he had 7 accounts and one was banned after forwarding without flood alerts. Please ask your friend if:
- has created a fresh new account
- has used a VoIP number
- he used the account right away to forward many msgs

Other users with banned accounts were in the same situation. I think that TG closely monitors newly created accounts and bans them for activities out of the ordinary. 

Usual a normal user makes an account and over the coming months, he subscribes to other channels, writes messages, and forwards from time to time. But forwarding thousand of messages from a newly created account is out of the ordinary for the majority of users. An that triggers some alerts on TG. And if a VoIP number is used is even worst, because TG keeps a closed eye on them, being largely used for questionable activities on TG.

Ask your friend to write to @spamBot and to be sincere. There's a chance that TG understands that is not engaged in bad activities.

I'm glad that you confirmed that the delay works as expected. And yes, besides flood msg may be other factors that lead to account bans. Like the ones, I mentioned before.

**Reply by ankit-sinha-18** [reply]
He had 7 accounts and all get blocked ,
Two were VOIP and rest were genuine number.

Btw, I have also seen that newly created accounts get blocked instantly.
His 5 TG account are created through genuine number but that was created on the same day,

I am also using VOIP number and I am using this number since 5 months.,
And from the the last month,I am using Telegram chat forward (tgcf) with VOIP number and I have forwarded millions of messages but still Telegram didn't ban my VOIP number.

So ,don't try to forward message with newly created Telegram account,it may lead to suspension of account if limit will cross 4000,6000 message onwards (experienced by me)

**Reply by tissole** [reply]
WOW! 7 accounts banned in 2 days! So 5 of those were created on the same day and used for large forwards. What about the rest of the 2 accounts? Were those old or also newly created?

A pattern clearly emerges, new accounts used for big forwards get an instant ban. Probably the TG system confuses large requests to some bad activities. Some guys sell scripts that scrape TG for users and forcibly add them to channels. And people who use those scripts don't make accounts and wait for weeks and months. They used them instantly rotating VoIP numbers. So large requests from newly created accounts look very suspicious to TG.

How many hours did you use this script uninterrupted?

**Reply by ankit-sinha-18** [reply]
Rest two were VOIP number ,
And Telegram account created on the same day with those VOIP numbers

And all these accounts get banned by forwarding only 4000-5000 message from each Telegram account

**Reply by tissole** [reply]
Thank you for the info. Is pretty clear now, a new account should not be used for big forwards. Users should use old accounts or wait for few weeks and slowly increasing their forwards.

P.S. You personally, how many hours did you use this script without pause and no ban? What was the longest period of time?

**Reply by ankit-sinha-18** [reply]
30 days approx . And still forwarding 24*7 hours.. 😊

**Reply by tissole** [reply]
OMG! Where you hosted the app? Currently, I'm thinking of repli.it. I started reading their documentation. Aaa..I see you use live mode, I need it for past mode, some big channels I want to clone.

**Reply by ankit-sinha-18** [reply]
No,I have used tgcf for past mode only.
I rarely use live feature.
I have hosted it on Heroku.

**Reply by tissole** [reply]
But how you handle the lack of persistency for offset value?

**Reply by ankit-sinha-18** [reply]
Just as you already suggested to someone to find out the last message from the source Channel, I also used this sometimes

But ,I generally check logs within a day before the dyno stops, In the logs I can see the message forwarding Id ,just stop the app and keep in mind that offset
And just update the offset id through config and restart the app again.

**Reply by tissole** [reply]
Ha...ha...this trick works, but kinda breaks the idea of automatic flow upon which this script is based. I think that a DB will overcome the limitations of non-persistent storage. It offers other advantages besides, preventing duplicates to be forward.

**Reply by ankit-sinha-18** [reply]
Create a discussion for implementing this feature 😁

**Reply by tissole** [reply]
That's a good idea, I will open a discussion on this subject.

**Comment by u-8d** [comment]
where you have deployed tgcf

**Reply by tissole** [reply]
I run it from my PC.

**Comment by tissole** [comment]
A little update. I run more tests. First with a delay of 1.60, after 1h 1962 msgs were forwarded and a pause of 262s was imposed by TG. So this might be the unsafe lower limit between delays. I didn't run it further.

Then I used a 1.61s delay and run for 4h. After the first hour a small pause of 108s was imposed but nothing for the rest. The forwarding rate was 1995 msg/h. This is probably safe but on the edge, better using 1.62 or 1.63. Now we need confirmation for longer timeframes 10h+.

**Comment by ghost** [comment]
It's very interesting. Thank you. I'll test this over the long term.

**Comment by warc0cara** [comment]
good night

 I'm from Brazil this telegram bot project (tgcf) is incredible, congratulations, I would like to know if I can delay receiving the forwarded message.
